apm-agent-plugins/apm-vertx/apm-vertx-common/src/main/java/co/elastic/apm/agent/vertx/AbstractRoutingContextHandlerWrapper.java

```diff
@@ -0,0 +1,59 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2021 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.vertx;
+
+import co.elastic.apm.agent.impl.transaction.Transaction;
+import io.vertx.core.Handler;
+import io.vertx.ext.web.RoutingContext;
+
+public abstract class AbstractRoutingContextHandlerWrapper implements Handler<RoutingContext> {
+
+    private final Handler<RoutingContext> actualHandler;
+
+    protected AbstractRoutingContextHandlerWrapper(Handler<RoutingContext> handler) {
```

Comment:
I think the way the routing context handlers are wrapped also doesn't work well with runtime attachment as the instrumentation is applied only when the routes are initialized.

A fairly easy workaround is to instrument `io.vertx.ext.web.impl.RouterImpl#handle` which is stable across versions. As an added benefit, the request is not wrapped, so you don't need to use magic to unwrap it.

Reply:
io.vertx.ext.web.impl.RouterImpl#handle is a bad place, because there the RoutingContext does not exist yet. But resolved it through instrumenting RouteImpl#handleContext / RouteState#handleContext